crates/communication/src/server/parameters/subscriptions.rs

```diff
@@ -0,0 +1,1359 @@
+use std::{
+    collections::{hash_map::Entry, BTreeMap, HashMap},
+    sync::Arc,
+};
+
+use framework::Reader;
+use futures_util::{stream::FuturesUnordered, StreamExt};
+use log::error;
+use serialize_hierarchy::{Format, SerializeHierarchy, SerializedValue};
+use tokio::{
+    select, spawn,
+    sync::{
+        mpsc::{Receiver, Sender},
+        Notify,
+    },
+    task::JoinHandle,
+};
+
+use crate::{
+    messages::{ParametersRequest, ParametersResponse, Path, Response, TextualResponse},
+    server::{get_paths_from_hierarchy, Client},
+};
+
+use super::{ClientRequest, StorageRequest};
+
+pub(crate) fn subscriptions<Parameters>(
+    mut request_receiver: Receiver<ClientRequest>,
+    parameters_reader: Reader<Parameters>,
+    parameters_changed: Arc<Notify>,
+    storage_request_sender: Sender<StorageRequest>,
+) -> JoinHandle<()>
+where
+    Parameters: Send + SerializeHierarchy + Sync + 'static,
+{
+    spawn(async move {
+        let fields = get_paths_from_hierarchy(Default::default(), Parameters::get_hierarchy());
+
+        let mut subscriptions = HashMap::new();
+        loop {
+            select! {
+                request = request_receiver.recv() => {
+                    let Some(request) = request else {
+                        break;
+                    };
+                    handle_request(
+                        request,
+                        &parameters_reader,
+                        &storage_request_sender,
+                        &mut subscriptions,
+                        &fields,
+                    ).await;
+                },
+                _ = parameters_changed.notified() => {
+                    handle_changed_parameters(&parameters_reader, &subscriptions).await;
+                }
+            }
+        }
+    })
+}
+
+async fn handle_request<Parameters>(
```

Comment:
This function has a large match statement where every branch (except one) tries to send at least one message.
I think the readability of this function would be aided greatly by moving all the send calls into once place, perhaps even returning `Option` and sending the message in an outside function or at least outside the match statement.

Sadly it's not quite so simple since it's two message types being sent to two different channels, but even so I think there is some potential for improvement here.

Reply:
Hopefully better now. Let's not overengineer it.